sets.c

```diff
@@ -0,0 +1,289 @@
+
+    This library is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published
+    by the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this library.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include <sys/types.h>
+#include <ctype.h>
+#include <assert.h>
+#include <errno.h>
+
+#include "show_message.h"
+#include "sets.h"
+
+#define DEFAULT_UINTLS_SZ (1000)
+
+/*
+ * -= The Set Data-Type =-
+ *
+ * A 'set' is a type similar in vein to the mathematical definition of a set.
+ * See: https://foldoc.org/set
+ *
+ * Sets support ranges. See the `delete` longdoc for more information.
+ *
+ * Restrictions of our set data-type:
+ * - No empty sets.
+ * - No duplicate elements.
+ * - All the elements must be of the same type.
+ */
+
+/*
+ * Used with `qsort()`.
+ *
+ * This function uses the less elegant branching solution
+ * to deal with possible overflow.
+ */
+static int _size_t_cmp(const void *_i1, const void *_i2)
+{
+    const size_t i1 = *(const size_t *)_i1, i2 = *(const size_t *)_i2;
+
+    if      (i1 <  i2) return -1;
+    else if (i1 == i2) return  0;
+    else               return  1;
+}
+
+bool parse_uintset(const char *lptr, struct set *set, size_t maxsz)
+{
+    const char        *tok, *tmpnum = NULL, *tmpnum_end;
+    char              *tmpnum_endptr = NULL, *fail_reason = "BUG";
+    bool               got_num, is_hex = false;
+    size_t             last_num = 0, arr_szfilled, arr_maxsz;
+    size_t            *valarr, *tmp_vaptr;
+    size_t             vaidx = 0;
+    unsigned long long toknum;
+
+    assert(lptr && set);
+
+    /* ensure set is initialized */
+    memset(set, 0, sizeof(*set));
+
+    enum {
+        NIL = -1,
+        NUMBER_TOK,
+        RANGE_TOK,
+        COMMA_TOK,
+    } last_type = NIL;
+
+    /* allocate space for value pointers */
+    if ((valarr = calloc(DEFAULT_UINTLS_SZ, sizeof(size_t))) == NULL) {
+        show_error("%s(): OOM (Out Of Memory)\n", __func__);
+        return false;
+    }
+
+    arr_szfilled = 0;
+    arr_maxsz    = DEFAULT_UINTLS_SZ;
+    got_num      = false;
+
+/*
+ * NOTE: used locally. fragile, make sure to not use it with anything fancy.
+ *
+ * We must use a separate pointer with realloc(), in case it fails. When that
+ * happens, valarr is _not_ free()'d by realloc(), and valarr then points to
+ * NULL.
+ */
+#define INC_ARRAY_SZ                                                                           \
```

Comment:
This thing keeps getting more and more complicated, it's probably time to make it a `static inline` function, so you can define local variables freely and make the structure clear.

Reply:
Changed.